### PR TITLE
BUILD: refactor circleci to use spin [skip actions][skip azp][skip cirrus]

### PR DESCRIPTION
**Pull request by mattip**
Use `spin` in circle CI. This is a small cleanup that came out of the triage discussion around doctesting, motivated by @ev-br. It also adds `scipy-doctest` to the doc building requirements, so developers can use `spin check-docs` to run the doctests via `scipy-doctest`.

### CODE REVIEW

**Comment by ev-br** [comment]
The refguide-check you're removing now was still doing something:

- checking that the routine listing in `__init__.py` files agree with `__all__` lists (the original refguide checking)
- doing some  ReST validation

It's not my call to make whether these checks are still wanted or not :-).

**Comment by mattip** [comment]
Hmm. We never activated the `--rst` check option in CI, so it was not really used. Is there an option in scipy-doctest to do RST block doctests?

**Comment by ev-br** [comment]
> Is there an option in scipy-doctest to do RST block doctests?

`spin check-tutorials` check _doctests_ in rst files under `numpy/doc/source/user` if that's what you mean. 
But rst syntax, no, it does not check any.
(if anything, the syntax checks sound more like `numpydoc` material maybe?)

**Comment by mattip** [comment]
In your comment about removing refguide-check, you refered to "ReST validation". In this change I did
```diff
- python tools/refguide_check.py -v
+ spin check-docs -v
+ spin check-tutorials -v
```

Is there some ReST validation the old code does that is removed in this change?

As for the `__init__` / `__all__` check: maybe we should add that to [test_public_api.py](https://github.com/numpy/numpy/blob/main/numpy/tests/test_public_api.py) since it is not really a documentation issue.

**Comment by ev-br** [comment]
> Is there some ReST validation the old code does that is removed in this change?

Yes. https://github.com/numpy/numpy/blob/main/tools/refguide_check.py#L392
_I think_  (not 100% sure) this basically whitelists ReST directives.
Whether it's useful to keep, I don't know.

> As for the __init__ / __all__ check: maybe we should add that to test_public_api.py since it is not really a documentation issue.

May add there of course. Or remove.
But this _is_ a documentation issue: basically, https://github.com/numpy/numpy/blob/main/tools/refguide_check.py#L318 checks that a listing like https://github.com/numpy/numpy/blob/main/numpy/fft/__init__.py indeed has all public objects.

**Comment by mattip** [comment]
Thanks for the clarifications. I put the call to `refguide_check` back and added some comments.

**Comment by ngoldbaum** [comment]
@mattip let me know if you want a hand getting this over the line

**Comment by mattip** [comment]
I'm not sure what the best path forward is here. Summary: scipy-doctest is not a complete replacement for refguide_check since the former is missing some capabilities. We could:
- Use refguide_check underneath `spin check-docs` instead of scipy-doctest 
- Use spin only for the build, and leave the doctest checks as they were
- Add the `__init__`/`__all__` and ReST validation checks to scipy-doctest and wait for a new release, then remove `refguide_check.py`

I guess I would prefer the last option, @ev-br is that something that is possible?

**Comment by ev-br** [comment]
TL;DR I think it's best to keep the current state with three separate calls on CI, `check-docs`, `check-tutorials` and `refguide-check`.

Indeed, scipy-doctest is not a complete replacement for refguide-check. refguide-check previously did three separate things, scipy-doctest does one of them, doctests only. Parsing full docstrings for ReST admonitions etc is out of scope for scipy-doctest, and so is analyzing submodule documentation for `__init__ / __all__` consistency. 

The two remaining tasks of refguide-check can be either kept as is IMO, or maybe outsourced to some external tools (esp ReST validation). 

AFAICS, the current CircleCI state is that `spin check-docs` runs fine, and `spin check-tutorials` fails with
`ERROR: file or directory not found: /home/circleci/repo/.spin/../../numpy/doc/source/user`  from https://github.com/numpy/numpy/blob/main/.spin/cmds.py#L398 . This seems to be the correct location for a local checkout, and seems to be working OK on main. So is there something special about how CircleCI checks out the repo?

**Comment by mattip** [comment]
> TL;DR I think it's best to keep the current state with three separate calls on CI, check-docs, check-tutorials and refguide-check.

OK, rather than remove refguide_check, I will trim it to only do the two missing tests, and not the complete doctesting that is done in scipy-doctest.

**Comment by ev-br** [comment]
> rather than remove refguide_check, I will trim it to only do the two missing tests, and not the complete doctesting that is done in scipy-doctest.

Agreed! 
This, I believe, is the current state of refguide-check.py.

**Comment by mattip** [comment]
CI is passing, including running the [refguide_check tests](https://app.circleci.com/pipelines/github/numpy/numpy/28692/workflows/5f7773e3-096e-4bd5-9d97-2e82a58a93e1/jobs/42505).

**Comment by ngoldbaum** [comment]
Awesome, thanks for carrying this through Matti!
